### PR TITLE
Check if intake is already extended in climb command

### LINKED ISSUE #78: Check if intake is already extended in climb command
If it is already extended, skip the delay.

### PR DESCRIPTION
**Pull request by MatthewZ0823**
Fix #78 

If the intake is already extended then the delay is skipped

### CODE REVIEW

File: src/main/java/frc/robot/commands/climber/Climb.java [hunk 1]
```diff
@@ -35,13 +36,17 @@ public Climb(DoubleSupplier speed, Climber climber, Intake intake) {
 
   @Override
   public void initialize() {
-    m_timer.start();
-    m_intake.setExtended(true);
+    if (m_intake.isExtended()) {
+      m_isIntakeExtended = true;
```
Comment (Jacob4649) [inline src/main/java/frc/robot/commands/climber/Climb.java]: extend cllimber piston here

File: src/main/java/frc/robot/commands/climber/Climb.java [hunk 2]
```diff
@@ -35,13 +36,17 @@ public Climb(DoubleSupplier speed, Climber climber, Intake intake) {
 
   @Override
   public void initialize() {
-    m_timer.start();
-    m_intake.setExtended(true);
+    if (m_intake.isExtended()) {
```
Comment (Jacob4649) [inline src/main/java/frc/robot/commands/climber/Climb.java:41]: `m_isIntakeExtended = m_intake.isExtended()`

File: src/main/java/frc/robot/commands/climber/Climb.java [hunk 3]
```diff
@@ -35,13 +36,17 @@ public Climb(DoubleSupplier speed, Climber climber, Intake intake) {
...
+    } else {
+      m_timer.start();
+      m_intake.setExtended(true);
+    }
   }
 
   @Override
   public void execute() {
-    if (m_intake.isExtended()) {
+    if (m_isIntakeExtended) {
```
Comment (Jacob4649) [inline src/main/java/frc/robot/commands/climber/Climb.java]: '
if (m_isIntakeExtended) {
  m_climber.set(m_speed.getAsDouble());
} else if (m_timer.get() > 0.5) {
  m_climber.setPiston(true);
  m_climber.set(m_speed.getAsDouble);
} else {
  m_climber.stop();
}
'

**Review by Jacob4649 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/climber/Climb.java (hunk 1 above)
Comment (MatthewZ0823) [reply src/main/java/frc/robot/commands/climber/Climb.java]: Resolved in commit 408960c

File: src/main/java/frc/robot/commands/climber/Climb.java (hunk 2 above)
Comment (MatthewZ0823) [reply src/main/java/frc/robot/commands/climber/Climb.java:41]: Resolved in commit 408960c

File: src/main/java/frc/robot/commands/climber/Climb.java (hunk 3 above)
Comment (MatthewZ0823) [reply src/main/java/frc/robot/commands/climber/Climb.java]: Resolved in commit 408960c

File: src/main/java/frc/robot/commands/climber/Climb.java (hunk 2 above)
Comment (Jacob4649) [reply src/main/java/frc/robot/commands/climber/Climb.java:41]: **

**Review by Jacob4649 — CHANGES_REQUESTED** [review]

**Review by Jacob4649 — APPROVED** [review]

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 4]
```diff
@@ -42,6 +42,10 @@ public void setExtended(Value value) {
     m_solenoid.set(value);
   }
 
+  public boolean isExtended() {
+    return m_solenoid.get() == Value.kForward ? true : false;
+  }
+
```
Comment (matthew-reynolds) [inline src/main/java/frc/robot/subsystems/Intake.java:48]: This is redundant, the ternary operator should be removed.

**Review by matthew-reynolds — COMMENTED** [review]
I don't think this will work as intended. We want the following behaviour:

 1. If the intake is extended, skip to step 3 immediately.
 2. Extend the intake and wait 0.5 sec for it to fully extend. Recall that `m_intake.isExtended()` will return `true` immediately after you start to extend, even though the piston will not yet be fully extended. This is why we need the delay.
3. Extend the climber pistons. Wait a short period to allow the climber to flip up.
4. Extend the climber using the DoubleSupplier.


I would propose:

 - Create an enum to track the 3 states of the command: `INTAKE_EXTENDING`, `CLIMBER_ENGAGING`, `CLIMBER_RUNNING` or something like that.

**In `initialize()`:**
 - Start the timer.
 - If intake is not extended, extend the intake, and set the state to 1 (`INTAKE_EXTENDING`).
 - Otherwise, set the state to `CLIMBER_ENGAGING`.

**In `execute()`:**
 - If the state is `INTAKE_EXTENDING` AND the required intake piston time has elapsed, set the state to `CLIMBER_ENGAGING` and reset the timer.
 - If the state is `CLIMBER_ENGAGING` AND the required climber piston time has elapsed, set the state to `CLIMBER_RUNNING`
 - If the state is `CLIMBER_RUNNING`, run the climber motor using the `DoubleProvider`.

Please open a follow-up PR to address these changes.

**Comment by matthew-reynolds** [comment]
#81 fixes this somewhat, but further changes are still required.

**Pull request merged**